service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java

```diff
@@ -150,139 +176,136 @@ public void reset(final PolicyEvaluationRequestCacheKey key) {
+            final BiFunction<String, String, String> getKey) {
+        String key = getKey.apply(zoneId, entityId);
         String timestamp = timestampValue();
+        logSetEntityTimestampsDebugMessage(timestamp, key, entityId, entityType);
+        set(key, timestamp);
+    }
+
+    private void setEntityIfNotExists(final String zoneId, final String entityId,
+            final BiFunction<String, String, String> getKey) {
+        String key = getKey.apply(zoneId, entityId);
+        setIfNotExists(key, timestampValue());
+    }
+
+    private void logSetEntityTimestampsDebugMessage(final String timestamp, final String key, final String entityId,
+            final EntityType entityType) {
         if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug(String.format("Setting timestamp for poliy set '%s'; key: '%s', value: '%s' ", policySetId,
-                    key, timestamp));
+            LOGGER.debug("Setting timestamp for {} '{}'; key: '{}', value: '{}'", entityType, entityId, key, timestamp);
         }
-        set(key, timestamp);
+    }
+
+    @Override
+    public void resetForPolicySet(final String zoneId, final String policySetId) {
+        resetForEntity(zoneId, policySetId, EntityType.POLICY_SET, AbstractPolicyEvaluationCache::policySetKey);
+    }
+
+    @Override
+    public void setPolicySetIfNotExists(final String zoneId, final String policySetId) {
+        setEntityIfNotExists(zoneId, policySetId, AbstractPolicyEvaluationCache::policySetKey);
     }
 
     @Override
     public void resetForResource(final String zoneId, final String resourceId) {
-        String timestamp = timestampValue();
-        Set<String> toKeys = getResourceTranslations(resourceTranslationKey(zoneId, resourceId));
-        Map<String, String> map = new HashMap<>();
-        for (String toKey : toKeys) {
-            map.put(toKey, timestamp);
-            logSetResourceTimestampsDebugMessage(timestamp, toKey, resourceId);
-        }
-        multiSet(map);
+        resetForEntity(zoneId, resourceId, EntityType.RESOURCE, AbstractPolicyEvaluationCache::resourceKey);
+    }
+
+    @Override
+    public void setResourceIfNotExists(final String zoneId, final String resourceId) {
+        setEntityIfNotExists(zoneId, resourceId, AbstractPolicyEvaluationCache::resourceKey);
     }
 
     @Override
     public void resetForResourcesByIds(final String zoneId, final Set<String> resourceIds) {
-        multisetForResources(zoneId, resourceIds.stream().collect(Collectors.toList()));
+        multisetForResources(zoneId, new ArrayList<>(resourceIds));
     }
 
     @Override
     public void resetForResources(final String zoneId, final List<ResourceEntity> resourceEntities) {
-        multisetForResources(zoneId, resourceEntities.stream().map(resource -> resource.getResourceIdentifier())
+        multisetForResources(zoneId, resourceEntities.stream().map(ResourceEntity::getResourceIdentifier)
                 .collect(Collectors.toList()));
     }
```

Comment:
We are collecting resource Ids into a list and then build a map of keys inside multisetForResources.  Could we combine it all in one call? Map streamed resourceEntities to resourceKey and collect into a map with the timestamp logging debug message for each... :-)